File: app/src/main/java/com/example/softmusic/playMusic/MusicPlayFragment.java
package com.example.softmusic.playMusic;

import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.os.Bundle;
import android.os.SystemClock;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import android.widget.Toast;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;

import com.example.softmusic.MainActivity;
import com.example.softmusic.R;
import com.example.softmusic.databinding.FragmentMusicPlayBinding;

import java.util.List;
import java.util.Objects;


public class MusicPlayFragment extends Fragment implements SeekBar.OnSeekBarChangeListener, View.OnClickListener {

    private MediaBrowserCompat mBrowser;
    private MediaControllerCompat mController;
    private final String TAG = "MediaPlayer";
    private FragmentMusicPlayBinding binding;

    private int nowPos = 0;
    private int lastPos = -1;


    public MusicPlayFragment(){

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBrowser = new MediaBrowserCompat(requireContext(),
                new ComponentName(requireContext(),MediaPlaybackService.class), //绑定服务
                mBrowserConnectionCallback, // 设置回调
                null);
    }

    @Override
    public void onStart() {
        super.onStart();
        mBrowser.connect();
    }

    @Override
    public void onStop() {
        super.onStop();
        mBrowser.disconnect();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_music_play, container, false);
        binding.seekBar.setOnSeekBarChangeListener(this);
        binding.playsong.setOnClickListener(this);
        binding.nextsong.setOnClickListener(this);
        binding.lastsong.setOnClickListener(this);
        return binding.getRoot();
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int i, boolean b) {

    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        int pos = seekBar.getProgress();
        nowPos = pos / 1000;
        lastPos = -1;
        mController.getTransportControls().seekTo(pos);
        mController.getTransportControls().play();
        binding.playsong.setBackgroundResource(R.drawable.outline_pause_24);
    }

    @SuppressLint({"SwitchIntDef", "NonConstantResourceId"})
    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.playsong) {
            if (mController != null) {
                switch (mController.getPlaybackState().getState()) {
                    case PlaybackStateCompat.STATE_PLAYING:
                        mController.getTransportControls().pause();
                        lastPos = nowPos;
                        break;
                    case PlaybackStateCompat.STATE_PAUSED:
                        mController.getTransportControls().play();
                        lastPos = -1;
                        break;
                    case PlaybackStateCompat.STATE_NONE:
                        mController.getTransportControls().play();
                        updateProgressThread mUpdateProgressThread = new updateProgressThread();
                        mUpdateProgressThread.start();
                        break;
                }

                Log.d(TAG, "onClick: " + nowPos);
                Log.d(TAG, "onClick: " + lastPos);
            }
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        ((MainActivity)requireActivity()).setTitle("Play");
    }

    // 连接状态的回调接口，连接成功的时候会调用这里的onConnected方法
    private final MediaBrowserCompat.ConnectionCallback mBrowserConnectionCallback =
            new MediaBrowserCompat.ConnectionCallback(){
        @Override
        public void onConnected() {
            super.onConnected();
            Log.d(TAG, "onConnected");
            if (mBrowser.isConnected()){
                // 如果连接成功 这里的mediaId就为MediaBrowserService.onGetRoot的返回值
                String mediaId = mBrowser.getRoot();
                mBrowser.unsubscribe(mediaId);
                mBrowser.subscribe(mediaId,
                        mBrowserSubscriptionCallback // 这里是订阅回调接口，当Service读取数据后会把数据发送回来，调用这个回调接口
                );
                mController = new MediaControllerCompat(requireContext(),mBrowser.getSessionToken());
                // 给Controller注册回调
                mController.registerCallback(mMediaControllerCallback);
            }

            MediaMetadataCompat metadataCompat = mController.getMetadata();
            updateDuration(metadataCompat);
        }

        @Override
        public void onConnectionSuspended() {
            super.onConnectionSuspended();
        }

        @Override
        public void onConnectionFailed() {
            super.onConnectionFailed();
        }
    };

    // 向媒体浏览器服务(MediaBrowserService)发起数据订阅请求的回调接口
    private final MediaBrowserCompat.SubscriptionCallback mBrowserSubscriptionCallback =
            new MediaBrowserCompat.SubscriptionCallback() {
        @Override
        public void onChildrenLoaded(@NonNull String parentId,
                                     @NonNull List<MediaBrowserCompat.MediaItem> children) {
            super.onChildrenLoaded(parentId, children);
            Log.d(TAG, "onChildrenLoaded");
            // children 就是Service返回回来的数据
            for (MediaBrowserCompat.MediaItem item:children){
                Log.d(TAG,
                        Objects.requireNonNull(item.getDescription().getTitle()).toString());
            }

            MediaMetadataCompat metadataCompat = mController.getMetadata();
            updateDuration(metadataCompat);
        }
    };

    // 这个是Controller 的回调，通过这个回调来更新UI
    private final MediaControllerCompat.Callback mMediaControllerCallback = new MediaControllerCompat.Callback() {
        @SuppressLint("SwitchIntDef")
        @Override
        public void onPlaybackStateChanged(PlaybackStateCompat state) {
            super.onPlaybackStateChanged(state);
            switch (state.getState()){
                case PlaybackStateCompat.STATE_NONE:
                    Toast.makeText(requireContext(),"none",Toast.LENGTH_LONG).show();
                    break;
                case PlaybackStateCompat.STATE_PAUSED:
                    binding.playsong.setBackgroundResource(R.drawable.outline_play_arrow_24);
                    Toast.makeText(requireContext(),"pause",Toast.LENGTH_LONG).show();
                    Log.d(TAG, "onPlaybackStateChanged: " + (int) state.getPosition());
                    break;
                case PlaybackStateCompat.STATE_PLAYING:
                    binding.playsong.setBackgroundResource(R.drawable.outline_pause_24);
                    Toast.makeText(requireContext(),"play",Toast.LENGTH_LONG).show();
                    Log.d(TAG, "onPlaybackStateChanged: " + (int) state.getPosition());
                    break;
            }
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            super.onMetadataChanged(metadata);
        }
    };

    private void updateDuration(MediaMetadataCompat metadataCompat){
        if (metadataCompat != null){
            int duration = (int) metadataCompat.getLong(MediaMetadataCompat.METADATA_KEY_DURATION);
            Log.d(TAG, "updateDuration: " + duration);
            if (duration > 0){
                binding.seekBar.setMax(duration);
            }
        } else {
            Log.d(TAG, "updateDuration: null" );
        }
    }

    private class updateProgressThread extends Thread{
        @Override
        public void run() {
            super.run();
            while (binding.seekBar.getProgress() < binding.seekBar.getMax()){
                if (nowPos == lastPos){
                    continue;
                }
                nowPos ++;
                binding.seekBar.setProgress(nowPos*1000);
                SystemClock.sleep(1000);
            }
        }
    }
}

File: app/src/main/java/com/example/softmusic/playMusic/MusicPlayFragment.kt
package com.example.softmusic.playMusic

import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.PagerSnapHelper
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.SCROLL_STATE_IDLE
import com.example.softmusic.MainActivity
import com.example.softmusic.MainViewModel
import com.example.softmusic.R
import com.example.softmusic.databinding.FragmentMusicPlayBinding
import com.example.softmusic.entity.MusicSong
import com.example.softmusic.entity.PlaylistSongCrossRef
import com.example.softmusic.room.DataBaseUtils
import com.example.softmusic.songList.ListBottomSheet
import java.text.SimpleDateFormat
import java.util.*


class MusicPlayFragment : Fragment(), SeekBar.OnSeekBarChangeListener, View.OnClickListener {

    private lateinit var musicPlayViewModel: MusicPlayViewModel
    private lateinit var mainViewModel: MainViewModel
    private var mController: MediaControllerCompat? = null
    private val TAG = "MusicPlayFragment"
    private lateinit var _binding: FragmentMusicPlayBinding
    private val binding get() = _binding
    private var repeatMode = MediaPlaybackService.DEFAULT

    private var currentPosition = 0


    private val adapter: MusicRecordAdapter by lazy {
        MusicRecordAdapter(listOf(), requireContext())
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        musicPlayViewModel = ViewModelProvider(requireActivity())[MusicPlayViewModel::class.java]
    }

    @SuppressLint("Recycle")
    @RequiresApi(Build.VERSION_CODES.P)
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        Log.d(TAG, "onCreateView")
        val dateFormat = SimpleDateFormat("mm:ss", Locale.CHINA)
        dateFormat.timeZone = TimeZone.getTimeZone("GMT+00:00")

        mainViewModel = (requireActivity() as MainActivity).mainViewModel
        repeatMode = mainViewModel.currentPlayMode.value!!
        if (mainViewModel.haveMusicFlag) {
            mController = (requireActivity() as MainActivity).mController
        }
        _binding = FragmentMusicPlayBinding.inflate(inflater, container, false)

        val layoutManager = LinearLayoutManager(requireContext())
        val pagerSnapHelper = PagerSnapHelper()
        pagerSnapHelper.attachToRecyclerView(binding.snapRecyclerview)

        layoutManager.orientation = LinearLayoutManager.HORIZONTAL
        binding.apply {
            seekBar.setOnSeekBarChangeListener(this@MusicPlayFragment)
            playsong.setOnClickListener(this@MusicPlayFragment)
            nextsong.setOnClickListener(this@MusicPlayFragment)
            lastsong.setOnClickListener(this@MusicPlayFragment)
            favoriteFlag.setOnClickListener(this@MusicPlayFragment)
            repeatMode.setOnClickListener(this@MusicPlayFragment)
            playList.setOnClickListener(this@MusicPlayFragment)
            snapRecyclerview.layoutManager = layoutManager
            snapRecyclerview.adapter = adapter
            snapRecyclerview.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                    when (newState) {
                        SCROLL_STATE_IDLE -> {
                            when {
                                layoutManager.findFirstCompletelyVisibleItemPosition() == currentPosition + 1 -> {
                                    binding.nextsong.performClick()
                                }
                                layoutManager.findFirstCompletelyVisibleItemPosition() == currentPosition - 1 -> {
                                    binding.lastsong.performClick()
                                }
                            }
                            currentPosition = layoutManager.findFirstCompletelyVisibleItemPosition()
                        }
                    }
                }
            })
            when(this@MusicPlayFragment.repeatMode){
                    MediaPlaybackService.DEFAULT -> {
                        repeatMode.setBackgroundResource(R.drawable.repeat_24px)
                    }
                    MediaPlaybackService.REPEAT_ONE -> {
                        binding.repeatMode.setBackgroundResource(R.drawable.repeat_one_24px)
                        adapter.setRecordList(listOf(mainViewModel.currentMusicId.value?.let { it1 ->
                            DataBaseUtils.getMusicSongById(
                                it1
                            )
                        }) as List<MusicSong>)
                    }
                    MediaPlaybackService.SHUFFLE -> {
                        binding.repeatMode.setImageResource(R.drawable.shuffle_24px)
                    }
                }
        }



        mainViewModel.run {
            currentMusicId.observe(viewLifecycleOwner) {
                currentPosition = mainViewModel.currentPlayList.value!!.indexOf(DataBaseUtils.getMusicSongById(it))
                binding.snapRecyclerview.scrollToPosition(currentPosition)
            }
            duration.observe(viewLifecycleOwner) {
                binding.seekBar.max = it
                binding.durationTime.text = dateFormat.format(Date(it.toLong()))
            }

            currentProgress.observe(viewLifecycleOwner) {
                binding.seekBar.progress = (it * 1000)
                binding.nowTime.text = dateFormat.format(Date((it * 1000).toLong()))
            }

            currentTitle.observe(viewLifecycleOwner) {
                binding.songTitle.text = it
            }

            currentArtist.observe(viewLifecycleOwner) {
                binding.artistName.text = it
            }

            currentPlayMode.observe(viewLifecycleOwner) {
            }

            currentPlayList.observe(viewLifecycleOwner) {
                currentPosition = mainViewModel.currentPlayList.value!!.indexOf(mainViewModel.currentMusicId.value?.let { it1 ->
                    DataBaseUtils.getMusicSongById(
                        it1
                    )
                })
                binding.snapRecyclerview.scrollToPosition(currentPosition)
                adapter.setRecordList(it)
            }

            likeFlag.observe(viewLifecycleOwner) {
                if (it == true) {
                    binding.favoriteFlag.setBackgroundResource(R.drawable.favorite_24px_yes)
                }
            }

            playbackState.observe(viewLifecycleOwner) {
                when (it.state) {
                    PlaybackStateCompat.STATE_PAUSED -> {
                        binding.playsong.setBackgroundResource(R.drawable.outline_play_arrow_64)
                    }
                    else -> {
                        binding.playsong.setBackgroundResource(R.drawable.outline_pause_64)
                    }
                }
            }

        }
        return binding.root
    }

    override fun onProgressChanged(p0: SeekBar?, p1: Int, p2: Boolean) {}

    override fun onStartTrackingTouch(p0: SeekBar?) {}

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        if (mainViewModel.haveMusicFlag) {
            val pos: Int = seekBar.progress
            mainViewModel.currentProgress.value = pos / 1000
            mainViewModel.lastProgress.value = -1
            mController?.transportControls?.seekTo(pos.toLong())
            mController?.transportControls?.play()
        }
    }

    @SuppressLint("SwitchIntDef", "NonConstantResourceId", "InflateParams")
    override fun onClick(view: View) {
        if (mainViewModel.haveMusicFlag) {
            when (view.id) {
                R.id.playsong -> {
                    when (mController?.playbackState?.state) {
                        PlaybackStateCompat.STATE_PLAYING -> {
                            mController!!.transportControls.pause()
                        }
                        PlaybackStateCompat.STATE_PAUSED -> {
                            mController!!.transportControls.play()
                        }
                        PlaybackStateCompat.STATE_NONE -> {
                            mController!!.transportControls.play()
                        }
                        PlaybackStateCompat.STATE_SKIPPING_TO_NEXT -> {
                            mController!!.transportControls.play()
                        }
                    }
                }
                R.id.nextsong -> {
                    mController?.transportControls?.skipToNext()
                }
                R.id.lastsong -> {
                    mController?.transportControls?.skipToPrevious()
                }
                R.id.favorite_flag -> {
                    if (!mainViewModel.allPlayListSongsCrossRef.value!!.contains(
                            PlaylistSongCrossRef(
                                mainViewModel.currentId.value!![1], mainViewModel.currentId.value!![0]
                            )
                        )
                    ) {
                        binding.favoriteFlag.setBackgroundResource(R.drawable.favorite_24px_yes)
                        DataBaseUtils.insertMusicSongRef(
                            PlaylistSongCrossRef(
                                mainViewModel.currentId.value!![1],
                                mainViewModel.currentId.value!![0]
                            )
                        )
                    }
                }
                R.id.repeat_mode -> {
                    val bundle = Bundle()

                    when (repeatMode) {
                        MediaPlaybackService.DEFAULT -> {
                            mainViewModel.currentPlayMode.value = MediaPlaybackService.SHUFFLE
                            repeatMode = MediaPlaybackService.SHUFFLE
                            binding.repeatMode.setBackgroundResource(R.drawable.shuffle_24px)
                        }
                        MediaPlaybackService.SHUFFLE -> {
                            mainViewModel.currentPlayMode.value = MediaPlaybackService.REPEAT_ONE
                            repeatMode = MediaPlaybackService.REPEAT_ONE
                            adapter.setRecordList(listOf(mainViewModel.currentMusicId.value?.let { it1 ->
                                DataBaseUtils.getMusicSongById(
                                    it1
                                )
                            }) as List<MusicSong>)
                            binding.repeatMode.setBackgroundResource(R.drawable.repeat_one_24px)

                        }
                        MediaPlaybackService.REPEAT_ONE -> {
                            mainViewModel.currentPlayMode.value = MediaPlaybackService.DEFAULT
                            repeatMode = MediaPlaybackService.DEFAULT
                            binding.repeatMode.setBackgroundResource(R.drawable.repeat_24px)
                        }
                    }
                    bundle.putInt("order", repeatMode)
                    mController?.transportControls?.sendCustomAction(
                        MediaPlaybackService.CHANGE_MODE,
                        bundle
                    )
                }
                R.id.play_list -> {
                    val listBottomSheet = ListBottomSheet()
                    listBottomSheet.show((requireActivity() as MainActivity).supportFragmentManager,ListBottomSheet.TAG)
                }
            }
        } else {
            Toast.makeText(requireContext(), "你还没有播放列表哦，去添加歌曲吧！", Toast.LENGTH_LONG).show()
        }
    }
}

